File: packages/js/node/auth-u2m/src/postgres.ts
import { createHash, randomUUID } from "node:crypto";
import type { Pool, PoolClient, QueryResultRow } from "pg";
import type { StorageAdapter } from "./runtime.ts";

const TABLE = "dbx_tools_auth_u2m_tokens";
const LOCK_RETRY_MILLIS = 50;

type PgPoolLike = Pick<Pool, "connect" | "query">;
type LockRow = QueryResultRow & { acquired: boolean };
type UnlockRow = QueryResultRow & { released: boolean };

function advisoryLockId(profile: string): bigint {
  return createHash("sha256").update(`dbx-tools-auth-u2m:${profile}`).digest().readBigInt64BE(0);
}

function isUndefinedTable(error: unknown): boolean {
  return typeof error === "object" && error !== null && "code" in error && error.code === "42P01";
}

function sleep(milliseconds: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, milliseconds));
}

/** Create a U2M storage adapter backed by a caller-owned `pg.Pool`. */
export function createStorage(pool: PgPoolLike): StorageAdapter {
  const leases = new Map<string, { client: PoolClient; lockId: bigint }>();

  return {
    async load(profile) {
      try {
        const result = await pool.query<{ token: unknown } & QueryResultRow>(
          `SELECT token FROM ${TABLE} WHERE profile = $1`,
          [profile],
        );
        const token = result.rows[0]?.token;
        return token === undefined ? undefined : JSON.stringify(token);
      } catch (error) {
        if (isUndefinedTable(error)) return undefined;
        throw error;
      }
    },

    async prepareWrite() {
      await pool.query(
        `CREATE TABLE IF NOT EXISTS ${TABLE} (profile TEXT PRIMARY KEY, token JSONB NOT NULL, updated_at TIMESTAMPTZ NOT NULL DEFAULT now())`,
      );
    },

    async save(profile, token) {
      await pool.query(
        `INSERT INTO ${TABLE}(profile, token) VALUES ($1, $2::jsonb) ON CONFLICT(profile) DO UPDATE SET token = EXCLUDED.token, updated_at = now()`,
        [profile, token],
      );
    },

    async remove(profile) {
      try {
        await pool.query(`DELETE FROM ${TABLE} WHERE profile = $1`, [profile]);
      } catch (error) {
        if (!isUndefinedTable(error)) throw error;
      }
    },

    async acquireLock(profile, timeoutMillis) {
      const client = await pool.connect();
      const lockId = advisoryLockId(profile);
      const deadline = Date.now() + Number(timeoutMillis);
      try {
        while (true) {
          const result = await client.query<LockRow>(
            "SELECT pg_try_advisory_lock($1::bigint) AS acquired",
            [lockId.toString()],
          );
          if (result.rows[0]?.acquired === true) {
            const lease = randomUUID();
            leases.set(lease, { client, lockId });
            return lease;
          }
          if (Date.now() >= deadline) throw new Error(`timed out locking profile ${profile}`);
          await sleep(LOCK_RETRY_MILLIS);
        }
      } catch (error) {
        client.release(error instanceof Error ? error : undefined);
        throw error;
      }
    },

    async releaseLock(lease) {
      const held = leases.get(lease);
      if (!held) throw new Error(`unknown Postgres storage lease ${lease}`);
      leases.delete(lease);
      let releaseError: Error | undefined;
      try {
        const result = await held.client.query<UnlockRow>(
          "SELECT pg_advisory_unlock($1::bigint) AS released",
          [held.lockId.toString()],
        );
        if (result.rows[0]?.released !== true) {
          throw new Error(`Postgres advisory lock ${held.lockId} was not held by this connection`);
        }
      } catch (error) {
        releaseError = error instanceof Error ? error : new Error(String(error));
        throw error;
      } finally {
        held.client.release(releaseError);
      }
    },

    name() {
      return "postgres";
    },
  };
}

File: packages/rs/auth-u2m-cli/src/postgres.rs
use std::time::Duration;

use async_trait::async_trait;
use dbx_tools_auth_u2m::{CredentialStore, Error, Result, StorageLock, Token};
use sha2::{Digest, Sha256};
use sqlx::{postgres::PgPoolOptions, PgPool};

pub struct PostgresStore {
    pool: PgPool,
}

impl PostgresStore {
    pub async fn connect(url: &str) -> Result<Self> {
        let pool = PgPoolOptions::new()
            .max_connections(5)
            .connect(url)
            .await
            .map_err(sqlx_error)?;
        Ok(Self { pool })
    }
}

struct PostgresLock {
    connection: sqlx::pool::PoolConnection<sqlx::Postgres>,
    lock_id: i64,
}

#[async_trait]
impl StorageLock for PostgresLock {
    async fn release(mut self: Box<Self>) -> Result<()> {
        sqlx::query("SELECT pg_advisory_unlock($1)")
            .bind(self.lock_id)
            .execute(&mut *self.connection)
            .await
            .map_err(sqlx_error)?;
        Ok(())
    }
}

#[async_trait]
impl CredentialStore for PostgresStore {
    async fn load(&self, profile: &str) -> Result<Option<Token>> {
        let raw: Option<serde_json::Value> = match sqlx::query_scalar(
            "SELECT token FROM dbx_tools_auth_u2m_tokens WHERE profile = $1",
        )
        .bind(profile)
        .fetch_optional(&self.pool)
        .await
        {
            Ok(raw) => raw,
            Err(error) if undefined_table(&error) => return Ok(None),
            Err(error) => return Err(sqlx_error(error)),
        };
        raw.map(serde_json::from_value)
            .transpose()
            .map_err(Into::into)
    }

    async fn prepare_write(&self) -> Result<()> {
        sqlx::query(
            "CREATE TABLE IF NOT EXISTS dbx_tools_auth_u2m_tokens (profile TEXT PRIMARY KEY, token JSONB NOT NULL, updated_at TIMESTAMPTZ NOT NULL DEFAULT now())",
        )
        .execute(&self.pool)
        .await
        .map_err(sqlx_error)?;
        Ok(())
    }

    async fn save(&self, profile: &str, token: &Token) -> Result<()> {
        sqlx::query(
            "INSERT INTO dbx_tools_auth_u2m_tokens(profile, token) VALUES ($1, $2) ON CONFLICT(profile) DO UPDATE SET token = EXCLUDED.token, updated_at = now()",
        )
        .bind(profile)
        .bind(serde_json::to_value(token)?)
        .execute(&self.pool)
        .await
        .map_err(sqlx_error)?;
        Ok(())
    }

    async fn delete(&self, profile: &str) -> Result<()> {
        let result = sqlx::query("DELETE FROM dbx_tools_auth_u2m_tokens WHERE profile = $1")
            .bind(profile)
            .execute(&self.pool)
            .await;
        if let Err(error) = result {
            if !undefined_table(&error) {
                return Err(sqlx_error(error));
            }
        }
        Ok(())
    }

    async fn lock(&self, profile: &str, timeout: Duration) -> Result<Box<dyn StorageLock>> {
        let mut connection = self.pool.acquire().await.map_err(sqlx_error)?;
        let lock_id = advisory_lock_id(profile);
        let deadline = tokio::time::Instant::now() + timeout;
        loop {
            let acquired: bool = sqlx::query_scalar("SELECT pg_try_advisory_lock($1)")
                .bind(lock_id)
                .fetch_one(&mut *connection)
                .await
                .map_err(sqlx_error)?;
            if acquired {
                return Ok(Box::new(PostgresLock {
                    connection,
                    lock_id,
                }));
            }
            if tokio::time::Instant::now() >= deadline {
                return Err(Error::LockTimeout(profile.to_owned()));
            }
            tokio::time::sleep(Duration::from_millis(50)).await;
        }
    }

    fn name(&self) -> &'static str {
        "postgres"
    }
}

fn advisory_lock_id(profile: &str) -> i64 {
    let digest = Sha256::digest(format!("dbx-tools-auth-u2m:{profile}").as_bytes());
    i64::from_be_bytes(digest[..8].try_into().expect("SHA-256 has eight bytes"))
}

fn sqlx_error(error: sqlx::Error) -> Error {
    Error::Storage(format!("Postgres: {error}"))
}

fn undefined_table(error: &sqlx::Error) -> bool {
    error
        .as_database_error()
        .and_then(|error| error.code())
        .is_some_and(|code| code == "42P01")
}
